idx/metric_definition.go

```diff
@@ -0,0 +1,352 @@
+	}
+
+	return bld.String()
+}
+
+// setMetricName interns the MetricName in an
+// object store and stores the addresses of those strings
+// in MetricName.nodes
+func (mn *MetricName) setMetricName(name string) {
+	nodes := strings.Split(name, ".")
+	mn.nodes = make([]uintptr, len(nodes))
+	for i, node := range nodes {
+		// TODO: add error checking? Fail somehow
+		nodePtr, err := IdxIntern.AddOrGet([]byte(node))
+		if err != nil {
+			log.Error(err)
+		}
+		mn.nodes[i] = nodePtr
+	}
+}
+
+// ExtensionType is required to use custom marshaling as an extension
+// with msgp
+func (mn *MetricName) ExtensionType() int8 {
+	return 95
+}
+
+// Len is required to use custom marshaling as an extension
+// with msgp
+func (mn *MetricName) Len() int {
+	return len(mn.String())
+}
+
+// MarshalBinaryTo is required to use custom marshaling as an extension
+// in msgp
+func (mn *MetricName) MarshalBinaryTo(b []byte) error {
+	copy(b, []byte(mn.String()))
+	return nil
+}
+
+// UnmarshalBinary is required to use custom marshaling as an extension
+// in msgp
+func (mn *MetricName) UnmarshalBinary(b []byte) error {
+	mn.setMetricName(string(b))
+	return nil
+}
+
+// TagKeyValue stores a Key/Value pair. The strings
+// are interned in an object store before they are assigned.
+type TagKeyValue struct {
+	Key   string
+	Value string
+}
+
+// String returns a Key/Value pair in the form of
+// 'key=value'
+func (t *TagKeyValue) String() string {
+	bld := strings.Builder{}
+
+	bld.WriteString(t.Key)
+	bld.WriteString("=")
+	bld.WriteString(t.Value)
+
+	return bld.String()
+}
+
+// TagKeyValues stores a slice of all of the Tag Key/Value pair combinations for a MetricDefinition
+type TagKeyValues []TagKeyValue
+
+// Strings returns a slice containing all of the Tag Key/Value pair combinations for a MetricDefinition.
+// Each item in the slice is in the form of 'key=value'
+func (t TagKeyValues) Strings() []string {
+	tags := make([]string, len(t))
+	for i, tag := range t {
+		tags[i] = tag.String()
+	}
+	return tags
+}
+
+// Helper functions to sort TagKeyValues
+func (t TagKeyValues) Len() int           { return len(t) }
+func (t TagKeyValues) Swap(i, j int)      { t[i], t[j] = t[j], t[i] }
+func (t TagKeyValues) Less(i, j int) bool { return t[i].Key < t[j].Key }
+
+// MetricDefinition stores information which identifies a single metric
+type MetricDefinition struct {
+	Id    schema.MKey
+	OrgId uint32
+	// using custom marshalling for MetricName
+	// if there is another way we should explore that
+	Name       MetricName `msg:"name,extension"`
+	Interval   int
+	Unit       string
+	mtype      mType
+	Tags       TagKeyValues
+	LastUpdate int64
+	Partition  int32
+}
+
+// NameWithTags returns a string version of the MetricDefinition's name with
+// all of its tagsin the form of 'name;key1=value1;key2=value2;key3=value3'
+func (md *MetricDefinition) NameWithTags() string {
+	bld := strings.Builder{}
+
+	md.Name.string(&bld)
+	sort.Sort(TagKeyValues(md.Tags))
```

Comment:
probably it would be better to sort at the time the tags get added

Reply:
The reason I added it there is to ensure it passes `TestTagSorting` in `memory_find_test.go`. Without it, it will fail. I do sort in `SetTags` and `SetId` also.

Reply:
In that case I guess that test doesn't make sense. Generally, if it's possible to guarantee that all the tags are getting sorted every time an MD gets created/modified, then i think this is preferable over sorting on read.